plugins/algoltek-usb/fu-algoltek-usb-device.c

```diff
@@ -0,0 +1,514 @@
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(GByteArray) version_data = NULL;
+
+	/* UsbDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_algoltek_usb_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	version_data = algoltek_device_rdv(self, error);
+	version_str = g_strdup_printf("%s", version_data->data);
+
+	fu_device_set_version(device, version_str);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usb_device_write_firmware(FuDevice *device,
+				      FuFirmware *firmware,
+				      FuProgress *progress,
+				      FwupdInstallFlags flags,
+				      GError **error)
+{
+	FuAlgoltekUsbDevice *self = FU_ALGOLTEK_USB_DEVICE(device);
+	g_autoptr(GBytes) blob_isp = NULL;
+	g_autoptr(GBytes) blob_payload = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 18, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 2, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 80, NULL);
+
+	if (!algoltek_device_en(self, error))
+		return FALSE;
+
+	if (!algoltek_device_rst(self, 0x200, error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 900);
+
+	if (!algoltek_device_wrr(self, 0x80AD, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80C0, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80C9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80D1, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80D9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80E1, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_wrr(self, 0x80E9, 0, error))
+		return FALSE;
+
+	if (!algoltek_device_rst(self, 0, error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 500);
+
+	/* get ISP image */
+	blob_isp = fu_firmware_get_image_by_id_bytes(firmware, "isp", error);
+	if (blob_isp == NULL)
+		return FALSE;
+	if (!algoltek_device_isp(self,
+				 blob_isp,
+				 AG_ISP_ADDR,
+				 fu_progress_get_child(progress),
+				 error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	if (!algoltek_device_bot(self, AG_ISP_ADDR, error))
+		return FALSE;
+
+	fu_device_sleep(FU_DEVICE(self), 1000);
+
+	if (!algoltek_device_ers(self, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	fu_device_sleep(FU_DEVICE(self), 500);
+
+	/* get payload image */
+	blob_payload = fu_firmware_get_image_by_id_bytes(firmware, FU_FIRMWARE_ID_PAYLOAD, error);
+	if (blob_payload == NULL)
+		return FALSE;
+	if (!algoltek_device_wrf(self, blob_payload, fu_progress_get_child(progress), error))
+		return FALSE;
+
+	fu_progress_step_done(progress);
+
+	if (!algoltek_device_rst(self, 0x100, error))
+		return FALSE;
+
+	/* success! */
+	return TRUE;
+}
+
+static void
+fu_algoltek_usb_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
```

Comment:
does the daemon give you a red warning with the values you can use? e.g. see https://github.com/fwupd/fwupd/wiki/Daemon-Warning:-FuProgress-steps and then with the correct values you can remove GUESSED.

Reply:
No, I don't see a red warning with the value. I still remove the `FU_PROGRESS_FLAG_GUESSED` flag.